xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,451 @@
+// better error handling, check `buffer_append_checked`
+pub fn buffer_append<BSIZE: u32, DSIZE: u32> (buffer: Buffer<BSIZE>, data: bits[DSIZE]) -> Buffer<BSIZE> {
+    if buffer_can_fit(buffer, data) == false {
+        trace_fmt!("Not enough space in the buffer! {} + {} <= {}", buffer.length, DSIZE, BSIZE);
+        fail!("not_enough_space", ());
+    } else {
+        ()
+    };
+
+    Buffer {
+        contents: (data as bits[BSIZE] << buffer.length) | buffer.contents,
+        length: DSIZE + buffer.length
+    }
+}
+
+#[test]
+fn test_buffer_append() {
+    let buffer = Buffer { contents: u32:0, length: u32:0 };
+    let buffer = buffer_append(buffer, u16:0xBEEF);
+    assert_eq(buffer, Buffer { contents: u32:0xBEEF, length: u32:16 });
+    let buffer = buffer_append(buffer, u16:0xDEAD);
+    assert_eq(buffer, Buffer { contents: u32:0xDEADBEEF, length: u32:32 });
+}
+
+// Appends `data` to a `buffer` if possible. Returns the status of the operation
+// and the updated buffer in a BufferResult structure
+pub fn buffer_append_checked<BSIZE: u32, DSIZE: u32> (buffer: Buffer<BSIZE>, data: bits[DSIZE]) -> BufferResult<BSIZE> {
+    if buffer_can_fit(buffer, data) == false {
+        BufferResult { status: BufferStatus::NO_ENOUGH_SPACE, buffer }
+    } else {
+        BufferResult {
+            status: BufferStatus::OK,
+            buffer: Buffer {
+                contents: (data as bits[BSIZE] << buffer.length) | buffer.contents,
+                length: DSIZE + buffer.length
+            }
+        }
+    }
+}
+
+#[test]
+fn test_buffer_append_checked() {
+    let buffer = Buffer { contents: u32:0, length: u32:0 };
+
+    let result1 = buffer_append_checked(buffer, u16:0xBEEF);
+    assert_eq(result1, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { contents: u32:0xBEEF, length: u32:16 }
+    });
+
+    let result2 = buffer_append_checked(result1.buffer, u16:0xDEAD);
+    assert_eq(result2, BufferResult {
+        status: BufferStatus::OK,
+        buffer: Buffer { contents: u32:0xDEADBEEF, length: u32:32 }
+    });
+
+    let result3 = buffer_append_checked(result2.buffer, u16:0xCAFE);
+    assert_eq(result3, BufferResult {
+        status: BufferStatus::NO_ENOUGH_SPACE,
+        buffer: result2.buffer
+    });
+}
+
+// Pops `length` amount of data from a `buffer`. Returns the modified buffer
+// and the popped data. The length of the data is the same as the buffer size.
+// It will fail if the buffer has no sufficient amount of data.
+// For calls that need better error handling, check `buffer_pop_checked`.
+pub fn buffer_pop<BSIZE: u32>(buffer: Buffer<BSIZE>, length: u32) -> (Buffer<BSIZE>, bits[BSIZE]) {
+    if buffer_has_at_least(buffer, length) == false {
+        trace_fmt!("Not enough data in the buffer!");
+        fail!("not_enough_data", ());
+    } else {
+        ()
+    };
+
+    let mask = (bits[BSIZE]:1 << length) - bits[BSIZE]:1;
```

Comment:
I think we could use bitslice instead.
```
buffer.contents[0:length]
```

Reply:
Same problem as in https://github.com/google/xls/pull/1167#discussion_r1388269110